Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/TrackTruthMatcher.cpp

```diff
@@ -0,0 +1,141 @@
+
+#include <map>
+#include <optional>
+
+namespace ActsExamples {
+
+TrackTruthMatcher::TrackTruthMatcher(const Config& config,
+                                     Acts::Logging::Level level)
+    : IAlgorithm("TrackTruthMatcher", level), m_cfg(config) {
+  if (m_cfg.inputTracks.empty()) {
+    throw std::invalid_argument("Missing input tracks");
+  }
+  if (m_cfg.inputParticles.empty()) {
+    throw std::invalid_argument("Missing input particles");
+  }
+  if (m_cfg.inputMeasurementParticlesMap.empty()) {
+    throw std::invalid_argument("Missing input measurement particles map");
+  }
+  if (m_cfg.outputTrackParticleMatching.empty()) {
+    throw std::invalid_argument("Missing output track particles matching");
+  }
+  if (m_cfg.outputParticleTrackMatching.empty()) {
+    throw std::invalid_argument("Missing output particle track matching");
+  }
+
+  m_inputTracks.initialize(m_cfg.inputTracks);
+  m_inputParticles.initialize(m_cfg.inputParticles);
+  m_inputMeasurementParticlesMap.initialize(m_cfg.inputMeasurementParticlesMap);
+  m_outputTrackParticleMatching.initialize(m_cfg.outputTrackParticleMatching);
+  m_outputParticleTrackMatching.initialize(m_cfg.outputParticleTrackMatching);
+}
+
+ActsExamples::ProcessCode TrackTruthMatcher::execute(
+    const ActsExamples::AlgorithmContext& ctx) const {
+  // Read input tracks
+  const auto& tracks = m_inputTracks(ctx);
+
+  // Read truth input collections
+  const auto& particles = m_inputParticles(ctx);
+  const auto& hitParticlesMap = m_inputMeasurementParticlesMap(ctx);
+
+  TrackParticleMatching trackParticleMatching;
+  ParticleTrackMatching particleTrackMatching;
+
+  // TODO this may be computed in a separate algorithm
+  // TODO can we wire this through?
+  std::map<SimBarcode, std::size_t> particleTruthHitCount;
+  for (const auto& [_, pid] : hitParticlesMap) {
+    particleTruthHitCount[pid]++;
+  }
+
+  // For each particle within a track, how many hits did it contribute
+  std::vector<ParticleHitCount> particleHitCounts;
+
+  for (const auto& track : tracks) {
+    // Get the majority truth particle to this track
+    identifyContributingParticles(hitParticlesMap, track, particleHitCounts);
+    if (particleHitCounts.empty()) {
+      ACTS_DEBUG(
+          "No truth particle associated with this trajectory with tip index = "
+          << track.tipIndex());
+      continue;
+    }
+
+    // Get the majority particleId and majority particle counts
+    // Note that the majority particle might be not in the truth seeds
+    // collection
+    ActsFatras::Barcode majorityParticleId =
+        particleHitCounts.front().particleId;
+    std::size_t nMajorityHits = particleHitCounts.front().hitCount;
+
+    if (particles.find(majorityParticleId) == particles.end()) {
+      ACTS_DEBUG(
+          "The majority particle is not in the input particle collection, "
+          "majorityParticleId = "
+          << majorityParticleId);
+      continue;
+    }
```

Comment:
this is the only place where we use the particle collection

is this necessary? should we really skip the matching if we do not find the particle?